core/iwasm/common/wasm_shared_memory.c

```diff
@@ -87,6 +92,53 @@ search_module(WASMModuleCommon *module)
     return NULL;
 }
 
+static void
+create_list_of_waiter_addresses(void *key, void *value, void *user_data)
+{
+    AtomicWaitAddressArgs *data = (AtomicWaitAddressArgs *)user_data;
+    if (data->len > 0) {
+        if (!(data->addr = wasm_runtime_realloc(data->addr,
+                                                data->len * sizeof(uint32)))) {
+            LOG_ERROR("failed to realloc atomic wait address list during "
+                      "wait_map traversal");
+            return;
+        }
+    }
+    else {
+        if (!(data->addr = wasm_runtime_malloc(sizeof(data->len) + 1))) {
+            LOG_ERROR("failed to malloc atomic wait address list during "
+                      "wait_map traversal");
+            return;
+        }
+    }
+    data->len++;
+    data->addr[data->len - 1] = key;
+}
+
+void
+notify_stale_threads_on_exception(WASMModuleInstanceCommon *module_inst)
+{
+    AtomicWaitAddressArgs *args = NULL;
+    if (!(args = wasm_runtime_malloc(sizeof(AtomicWaitAddressArgs)))) {
+        LOG_ERROR("failed to create atomic wait address args node");
+        return;
+    }
+
+    memset(args, 0, sizeof(*args));
+    os_mutex_lock(&shared_memory_list_lock);
+    // create list of addresses
+    bh_hash_map_traverse(wait_map, create_list_of_waiter_addresses, args);
```

Comment:
Why not traverse first time to get the total size of wait_map, use it to allocate memory for `AtomicWaitAddressArgs *args`  with size `offsetof(AtomicWaitAddressArgs, addr) + sizeof(void *) * total_size` , and then traverse again without allocating memory?

Reply:
just wondering if there is any specific advantage of traversing two times? is it to ensure the final traversal is safe i.e. we don't go out of bounds?

Reply:
The purpose of that is to avoid multiple reallocations - knowing the size in advance will let you make only one allocation.